### PR TITLE
(Delivery) Handle error when hard link for file already exists

### PR DESCRIPTION
**Pull request by diitaz93**
## Description

`cg deliver` raises an error if files are already linked for a case.

### Added

- Counter variable `number_previously_linked_files`.

### Changed

- The error type when no files are found to `MissingFilesError`.
- The name of the counter `number_linked_files` to `number_linked_files_now`.
- When the file already exists, increase the counter `number_previously_linked_files` by one after warning the user.
- The guard checks if both `number_previously_linked_files`. and `number_linked_files_now` is zero.

### How to prepare for test

- [x] Ssh to relevant server (depending on type of change)
- [x] Use stage: `us`
- [x] Paxa the environment: `paxa`
- [x] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_main -t cg -b delivery-patch -a
    ```

### How to test

- [x] run `cg deliver analysis -d fastq -c busyhorse` 

### Expected test outcome

- [x] Raise warnings about the existence of the files, state that no new files were linked but finish without raising errors
(See output below)

## Review

- [x] Tests executed by SD
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [x] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Deploy this branch on 
  - [ ] Stage
  - [ ] Production

### CODE REVIEW

File: cg/meta/deliver.py [hunk 1]
```diff
@@ -175,7 +175,7 @@ def deliver_sample_files(
         if not self.dry_run:
             delivery_base.mkdir(parents=True, exist_ok=True)
         file_path: Path
-        number_linked_files: int = 0
+        number_reviewed_files: int = 0
```
Comment (seallard) [inline cg/meta/deliver.py:178]: I think the original variable names was ok, we need to introduce one more:

```python
number_of_files_linked_now: int = 0
number_of_files_linked_previously: int = 0
```

File: cg/meta/deliver.py [hunk 2]
```diff
@@ -186,18 +186,21 @@ def deliver_sample_files(
             out_path: Path = delivery_base / file_name
             if self.dry_run:
                 LOG.info(f"Would hard link file {file_path} to {out_path}")
-                number_linked_files += 1
+                number_reviewed_files += 1
```
Comment (seallard) [inline cg/meta/deliver.py:189]: Increment the `number_of_files_linked_now` here instead

File: cg/meta/deliver.py [hunk 3]
```diff
@@ -186,18 +186,21 @@ def deliver_sample_files(
...
                 continue
             LOG.info(f"Hard link file {file_path} to {out_path}")
             try:
                 os.link(file_path, out_path)
-                number_linked_files += 1
+                number_reviewed_files += 1
```
Comment (seallard) [inline cg/meta/deliver.py:194]: Increment the number_of_files_linked_now here instead

File: cg/meta/deliver.py [hunk 4]
```diff
@@ -186,18 +186,21 @@ def deliver_sample_files(
...
             except FileExistsError:
-                LOG.info(f"Path {out_path} exists, skipping")
-        if number_linked_files == 0:
-            raise DeliveryReportError(f"No files were linked for sample {sample_name}")
+                LOG.info(
+                    f"Warning: Path {out_path} exists, no hard link was made for file {file_name}"
+                )
+                number_reviewed_files += 1
```
Comment (seallard) [inline cg/meta/deliver.py:199]: Increment the `number_of_files_linked_previously` here

File: cg/meta/deliver.py [hunk 5]
```diff
@@ -186,18 +186,21 @@ def deliver_sample_files(
...
+        if number_reviewed_files == 0:
```
Comment (seallard) [inline cg/meta/deliver.py:200]: Change this if statement to
```python
if number_of_files_linked_previously == 0 and number_of_files_linked_now == 0
```

File: cg/meta/deliver.py [hunk 6]
```diff
@@ -186,18 +186,21 @@ def deliver_sample_files(
...
+            raise MissingFilesError(f"No files were linked for sample {sample_name}")
 
-        LOG.info(f"Linked {number_linked_files} files for sample {sample_id}, case {case_id}")
+        LOG.info(f"Linked {number_reviewed_files} files for sample {sample_id}, case {case_id}")
```
Comment (seallard) [inline cg/meta/deliver.py:203]: Update the log message, either 
```python
LOG.info(f"Linked {number_of_files_linked_now} files for sample {sample_id}, case {case_id}")
```

Or more explicit
```python
previous_files_info = ""

if number_of_files_linked_previously > 0:
    previous_files_info = f", Previously linked files: {number_of_files_linked_previously}"

log_message = f"Linked {number_of_files_linked_now} files for sample {sample_id}, case {case_id {previous_files_info}"

LOG.info(log_message)
```

**Comment by diitaz93** [comment]
## Test negative control
### 1
```bash
[js.diazboada@hasta:~] [S_base] 3m0s $ cg deliver analysis -d fastq -c funtomcat
/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:2231: SAWarning: Cascade settings "delete, merge, save-update" should not be combined with a viewonly=True relationship.   This configuration will raise an error in version 1.4.  Note that in versions prior to 1.4, these cascade settings may still produce a mutating effect even though this relationship is marked as viewonly=True.
  "viewonly=True." % (", ".join(sorted(non_viewonly)))
Running CG deliver
Set dry run to False
/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:1998: SAWarning: Setting backref / back_populates on relationship Demux.samples to refer to viewonly relationship Sample.demuxes should include sync_backref=False set on the Demux.samples relationship.  (this warning may be suppressed after 10 occurrences)
  (rel_b, rel_a, rel_b),
/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:1998: SAWarning: Setting backref / back_populates on relationship Sample.demuxes to refer to viewonly relationship Demux.samples should include sync_backref=False set on the Sample.demuxes relationship.  (this warning may be suppressed after 10 occurrences)
  (rel_b, rel_a, rel_b),
Fetch latest version from bundle funtomcat
Setting ticket to 336161
Setting customer_id to cust002
Fetch latest version from bundle ACC11044A8
Traceback (most recent call last):
  File "/home/proj/stage/bin/miniconda3/envs/S_main/bin/cg", line 8, in <module>
    sys.exit(base())
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/decorators.py", line 27, in new_func
    return f(get_current_context().obj, *args, **kwargs)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/cg/cli/deliver/base.py", line 108, in deliver_analysis
    deliver_api.deliver_files(case_obj=case_obj)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/cg/meta/deliver.py", line 118, in deliver_files
    version_obj=last_version,
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/cg/meta/deliver.py", line 202, in deliver_sample_files
    raise MissingFilesError(f"No files were linked for sample {sample_name}")
cg.exc.MissingFilesError: No files were linked for sample 22418-I-1A
[js.diazboada@hasta:~] [S_base] 24s 1 $ 
```
### 2 
```bash
[js.diazboada@hasta:~] [S_base] 24s 1 $ cg deliver analysis -d fastq -t 336161
/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:2231: SAWarning: Cascade settings "delete, merge, save-update" should not be combined with a viewonly=True relationship.   This configuration will raise an error in version 1.4.  Note that in versions prior to 1.4, these cascade settings may still produce a mutating effect even though this relationship is marked as viewonly=True.
  "viewonly=True." % (", ".join(sorted(non_viewonly)))
Running CG deliver
Set dry run to False
/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:1998: SAWarning: Setting backref / back_populates on relationship Demux.samples to refer to viewonly relationship Sample.demuxes should include sync_backref=False set on the Demux.samples relationship.  (this warning may be suppressed after 10 occurrences)
  (rel_b, rel_a, rel_b),
/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/sqlalchemy/orm/relationships.py:1998: SAWarning: Setting backref / back_populates on relationship Sample.demuxes to refer to viewonly relationship Demux.samples should include sync_backref=False set on the Sample.demuxes relationship.  (this warning may be suppressed after 10 occurrences)
  (rel_b, rel_a, rel_b),
Fetch latest version from bundle equalcow
Setting ticket to 336161
Setting customer_id to cust002
Fetch latest version from bundle ADM1162A4
Traceback (most recent call last):
  File "/home/proj/stage/bin/miniconda3/envs/S_main/bin/cg", line 8, in <module>
    sys.exit(base())
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/click/decorators.py", line 27, in new_func
    return f(get_current_context().obj, *args, **kwargs)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/cg/cli/deliver/base.py", line 108, in deliver_analysis
    deliver_api.deliver_files(case_obj=case_obj)
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/cg/meta/deliver.py", line 118, in deliver_files
    version_obj=last_version,
  File "/home/proj/stage/bin/miniconda3/envs/S_main/lib/python3.7/site-packages/cg/meta/deliver.py", line 202, in deliver_sample_files
    raise MissingFilesError(f"No files were linked for sample {sample_name}")
cg.exc.MissingFilesError: No files were linked for sample 2015-14429
θ61° [js.diazboada@hasta:~] [S_base] 23s 1 $ 
```

**Comment by diitaz93** [comment]
@seallard done with changes and tests

**Review by henrikstranneheim — APPROVED** [review]
👍

File: cg/meta/deliver.py [hunk 7]
```diff
@@ -175,7 +175,8 @@ def deliver_sample_files(
         if not self.dry_run:
             delivery_base.mkdir(parents=True, exist_ok=True)
         file_path: Path
-        number_linked_files: int = 0
+        number_currently_linked_files: int = 0
```
Comment (seallard) [inline cg/meta/deliver.py:178]: I think this name is misleading, prefer number_of_files_linked_now

Comment (seallard) [reply cg/meta/deliver.py:178]: For all we know, the number of files linked currently is 100. But maybe there is a better name

Comment (diitaz93) [reply cg/meta/deliver.py:178]: true!

**Review by seallard — APPROVED** [review]

**Pull request merged**